### PR TITLE
Switch frame with link hints and refactor link hint modes

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR
* refactors the code for modes in `LinkHints` to
  - allow diffferent `linkActivators` and `getVisibleClickable` implementations for each mode
  - make it clearer which mode does what
  - make new mode definitions orthoganal to the definition of `LinkHintsMode`
* fixes bugs where `COPY_LINK_URL` and `OPEN_INCOGNITO` focus the `contentEditable` element containing a link instead of working on the link itself
* adds a new command (`activateModeToChangeFrame`) that selects frames using link hints
  - this is not yet exposed as a mappable command
  - when cross-frame link hints are implemented, this can be exposed to fix issue #494

This is the first part of an effort towards cross-frame link hints (#602), and also allows for useful possibilities such as #425. Each commit here is atomic, and so the commits may be easier to check than the PR diff.

### CODE REVIEW

**Comment by gdh1995** [comment]
Should we mark `<iframe>`s  when `f` is pressed, instead of inventing a new command to find frames specially? Just say I open an unfamiliar page and press `f`, then I might see nothing but a isolated HUD, but there's another kind of behavior:
1. Vimium shows a mark on the top left - like `[a]`
2. I press `a` and the iframe window gets focus
3. I press `f` again, now many marks will be shown properly.

**Comment by ekucks21** [comment]
+1 for getting this in

**Comment by smblott-github** [comment]
Switch frame with link hints sounds good, @mrmr1993 (if we can do it without a new command).

I'm not so keen on the refactoring.

**Comment by mrmr1993** [comment]
> I'm not so keen on the refactoring.

The main idea was to make each link hint mode's version of `getVisibleClickable` easily accessible (ie. without having to instantiate the full mode in every frame), so that we can call the correct one easily when doing cross-frame link hints.

Does this make the refactor any more acceptable/is there an alternative way that you'd like to see this done? Or should I just bin this?

**Comment by smblott-github** [comment]
Do we really need/want a new command for this?

I tried out a poor-man's version of the main idea here: 013965007858f1f80477a422c149b2b1695d2a2c.

It turns out it's not as useful as I'd expected, the problem being that - once you're in a frame - you cannot use link hints to get out of it.  That problem would be solved if we had global hints, but *that* feature would be extremely complex.

A simpler alternative might be:

- as currently, include all clickable things in the current frame,
- however, also include a single hint for each other frame which is in the current viewport.

Bottom line... I'm not sure we should do this in its current form.

(Entirely separately, it would be good if `gf` didn't select frames which are not in the viewport.)
